### PR TITLE
Add Test for TestRunAdmin.delete_view()

### PR DESCRIPTION
**Pull request by man4oni**

### CODE REVIEW

File: tcms/testruns/tests/test_admin.py [hunk 1]
```diff
@@ -14,6 +14,12 @@ def setUpTestData(cls):
         initiate_user_with_default_setups(cls.tester)
         cls.test_run = TestRunFactory()
 
+        cls.superuser = UserFactory()
+        cls.superuser.is_staff = True
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: is_staff is configured by UserFactory and is not needed

File: tcms/testruns/tests/test_admin.py [hunk 2]
```diff
@@ -22,3 +28,20 @@ def test_regular_user_can_not_change_testrun(self):
         response = self.client.get(reverse('admin:testruns_testrun_change',
                                            args=[self.test_run.pk]))
         self.assertRedirects(response, reverse('testruns-get', args=[self.test_run.pk]))
+
+    def test_regular_user_can_not_delete_testrun(self):
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertRaisesMessage(response, ('Permission denied: TestRun does not belong to you'))
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: assertRaisesMessage is used when you want to verify that a method/function call raises an Exception with a particular error message string. It doesn't apply here. 

The error message is shown as a warning on the web page and it is part of the HTML. You already know a way how to verify what is contained in the HTML so use it.

File: tcms/testruns/tests/test_admin.py [hunk 3]
```diff
@@ -22,3 +28,20 @@ def test_regular_user_can_not_change_testrun(self):
...
+        self.client.login(
+            username=self.superuser.username,
+            password='password')
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertContains(response, ("Yes, I'm sure"))
+
+        response = self.client.post(reverse('admin:testruns_testrun_delete',
+                                            args=[self.test_run.pk]), {'post': 'yes'})
+        self.assertRedirects(response, reverse('admin:testruns_testrun_changelist'))
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: This test method is generally OK but needs additional work:

- After deletion you need to verify that TR has actually been removed from the DB
- ATM I'm not sure if removing the test object introduces dependencies between the test methods and the order in which they are executed. I'll have to consult the docs to see how the transaction engine handles the deletion. The easiest way for you to check that is to have another test method which will execute after this one and will assert that this TR still exists in the DB. Django test runner sorts methods by name so something like `test_zzzz` should be executed after the current method.

File: tcms/testruns/tests/test_admin.py [hunk 4]
```diff
@@ -22,3 +28,20 @@ def test_regular_user_can_not_change_testrun(self):
...
+                                           args=[self.test_run.pk]))
+        self.assertRaisesMessage(response, ('Permission denied: TestRun does not belong to you'))
+
+    def test_superuser_can_delete_testrun(self):
+        self.client.login(
+            username=self.superuser.username,
+            password='password')
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertContains(response, ("Yes, I'm sure"))
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: This message is marked for translation so you will have to assert on the translated text as well. That's why we're testing with French & Slovenian translations and you get a failure for Slovenian. Search for the gettext_lazy function.

File: tcms/testruns/tests/test_admin.py [hunk 5]
```diff
@@ -22,3 +28,21 @@ def test_regular_user_can_not_change_testrun(self):
...
+        self.client.login(
+            username=self.superuser.username,
+            password='password')
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertContains(response, _("Yes, I'm sure"))
+        response = self.client.post(reverse('admin:testruns_testrun_delete',
+                                            args=[self.test_run.pk]), {'post': 'yes'}, follow=True)
+        self.assertRedirects(response, reverse('admin:testruns_testrun_changelist'))
+        self.assertNotContains(response, ('<a href="/admin/testruns/testrun/%d/change/">%s</a>'),
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: This is missing the parameters for string formatting and will PASS even if the delete fails.

A more accurate way to verify that a record is missing is to query the database directly. Use TestRun.objects.filter() and .exists() methods together with the TR ID which you already have to verify that the object we deleted is not found in the database.

File: tcms/testruns/tests/test_admin.py [hunk 6]
```diff
@@ -22,3 +29,23 @@ def test_regular_user_can_not_change_testrun(self):
...
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertContains(response, _("Yes, I'm sure"))
+        response = self.client.post(reverse('admin:testruns_testrun_delete',
+                                            args=[self.test_run.pk]), {'post': 'yes'}, follow=True)
+        self.assertRedirects(response, reverse('admin:testruns_testrun_changelist'))
+
+    def test_superuser_can_delete_testrun_verify(self):
+        trpk = self.test_run.pk
+        return TestRun.objects.filter(pk=trpk).exists()
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: These 2 lines can be condensed into 1 and need to be at the end of test_superuser_can_delete_testrun method, not separate. 

Also `return` here is not needed.

File: tcms/testruns/tests/test_admin.py [hunk 7]
```diff
@@ -22,3 +29,20 @@ def test_regular_user_can_not_change_testrun(self):
...
+        self.client.login(  # nosec:B106:hardcoded_password_funcarg
+            username=self.superuser.username,
+            password='password')
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertContains(response, _("Yes, I'm sure"))
+        response = self.client.post(reverse('admin:testruns_testrun_delete',
+                                            args=[self.test_run.pk]), {'post': 'yes'}, follow=True)
+        self.assertRedirects(response, reverse('admin:testruns_testrun_changelist'))
+        TestRun.objects.filter(pk=self.test_run.pk).exists()
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: Ahhh, what does this do here? 

You are asking the database for something and completely ignoring to validate the result. Regardless of the answer the test will always report PASS.

**Comment by zahari** [comment]
Monkey see monkey do :)

File: tcms/testruns/tests/test_admin.py [hunk 8]
```diff
@@ -22,3 +29,22 @@ def test_regular_user_can_not_change_testrun(self):
...
+            password='password')
+        print(TestRun.objects.values_list('pk'))
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertContains(response, _("Yes, I'm sure"))
+        response = self.client.post(reverse('admin:testruns_testrun_delete',
+                                            args=[self.test_run.pk]), {'post': 'yes'}, follow=True)
+        self.assertRedirects(response, reverse('admin:testruns_testrun_changelist'))
+        if TestRun.objects.filter(pk=self.test_run.pk).exists():
+            raise Exception('testruns delete unsuccessful')
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: You have assert methods to do exactly that.v

File: tcms/testruns/tests/test_admin.py [hunk 9]
```diff
@@ -22,3 +29,20 @@ def test_regular_user_can_not_change_testrun(self):
...
+        self.client.login(  # nosec:B106:hardcoded_password_funcarg
+            username=self.superuser.username,
+            password='password')
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertContains(response, _("Yes, I'm sure"))
+        response = self.client.post(reverse('admin:testruns_testrun_delete',
+                                            args=[self.test_run.pk]), {'post': 'yes'}, follow=True)
+        self.assertRedirects(response, reverse('admin:testruns_testrun_changelist'))
+        self.assertQuerysetEqual(TestRun.objects.values_list('pk'), [])
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py]: .filter().exists() returns a boolean. And you already know how to compare 2 values for equality. 

I prefer having .exists() here b/c it makes the assertion dead obvious.

File: tcms/testruns/tests/test_admin.py [hunk 10]
```diff
@@ -22,3 +29,20 @@ def test_regular_user_can_not_change_testrun(self):
...
+        self.client.login(  # nosec:B106:hardcoded_password_funcarg
+            username=self.superuser.username,
+            password='password')
+        response = self.client.get(reverse('admin:testruns_testrun_delete',
+                                           args=[self.test_run.pk]))
+        self.assertContains(response, _("Yes, I'm sure"))
+        response = self.client.post(reverse('admin:testruns_testrun_delete',
+                                            args=[self.test_run.pk]), {'post': 'yes'}, follow=True)
+        self.assertRedirects(response, reverse('admin:testruns_testrun_changelist'))
+        self.assertEqual(TestRun.objects.filter(pk=self.test_run.pk).exists(), False)
```
Comment (atodorov) [inline tcms/testruns/tests/test_admin.py:48]: This is acceptable. Another variant could be to use `assertFalse()`

**Comment by zahari** [comment]
Little by little tragedy is being improved to saga

**Pull request merged**